Gemfile

```diff
@@ -65,6 +65,9 @@ group :development do
 
   gem "spring"
   gem "spring-watcher-listen", "~> 2.0.0"
+
+  gem "rails-erd"
+  gem "ruby-graphviz"
```

Comment:
[Q]
`"ruby-graphviz"` は`"rails-erd"`の内部gemとして利用されているので、個別でインストールしなくても大丈夫だと思いますが、記述しないと動かなかった感じですかね？

### 参考
https://rubygems.org/gems/rails-erd

Reply:
こちら最初にgraphvizをインストールするべきと知り、試みましたがconflictを起こしたので先にrails-erdのみ導入したところ
""Post-install message from ruby-graphviz:
You need to install GraphViz (http://graphviz.org/) to use this Gem.""
と出たので、そこから調べて記述しました。

知識が乏しくて申し訳ないのですが、参考ページはどこをみれば『内部gem』だと判断できるのでしょうか？runtime依存性は『そのgemを使うのに必要なgem一覧』という認識だったのですが、もしかすると『必要なので個別にインストールしなくても良い』という意味も含まれているのでしょうか？

Reply:
なるほど！ご自身で調べられていてナイスです！もちろん、gem installでも間違いではないのですが、

```
""Post-install message from ruby-graphviz:
You need to install GraphViz (http://graphviz.org/) to use this Gem.""
```

と出ているので、単純にGraphVizをインストールするでも実はOKだったりします。

Macをお使いでしたら、Homebrewでインストールできますしね。

```
$ brew install graphviz
```

こんな感じで。

Reply:
>参考ページはどこをみれば『内部gem』だと判断できるのでしょうか？

ちょっと端折りすぎましたね💦

>runtime依存性は『そのgemを使うのに必要なgem一覧』という認識だった

その認識で間違いないですよ！

>参考ページはどこをみれば『内部gem』だと判断できるのでしょうか？

Runtime依存性の箇所ですね！

>『必要なので個別にインストールしなくても良い』という意味も含まれているのでしょうか？

こちらも大筋として上記に認識で問題ないかと！

ちょっと細かい話になってしまうのですが、rails-erdのリポジトリをみるとわかりますね。

プロジェクトルートに、 `rails-erd.gemspec` というファイルがありますよね。

## 参考
https://github.com/voormedia/rails-erd/blob/master/rails-erd.gemspec

これは、  `rails-erd` をgemとして作るときに必要な要件定義書みたいな物です。

その中に、  `add_runtime_dependency` という項目があり、そこで指定されたGemが `rails-erd` を使うのに必要なgemです。

いつもRailsでGemを使うときにはGemfileに追記しますが、 gemとして作る時だけは `.gemspec` というファイルに記述します。

結果、 `add_runtime_dependency`　に記載されたgemは、 railsなどでGemfileに `gem rails-erd` と記載した際に一緒にインストールされるため、その際個別にインストールする必要がなくなるというわけですね。

Reply:
内部 gem というか、rails-erd が ruby-graphviz に依存している（rails-erd 自体が ruby-graphviz を使っている）のは、実は Gemfile.lock を見るだけでわかったりします。参考まで。

https://github.com/Takahiro800/qiita_clone1/pull/15/files#diff-e79a60dc6b85309ae70a6ea8261eaf95R171

Reply:
>単純にGraphVizをインストールするでも実はOKだったりします。
>Macをお使いでしたら、Homebrewでインストールできますしね。
言葉足らずで申し訳ありません。
実はHome brewでインストールを試みた結果、以下のようなエラー？文が出ました
```==> Pouring graphviz-2.42.2.mojave.bottle.tar.gz
🍺  /usr/local/Cellar/graphviz/2.42.2: 508 files, 11MB
==> `brew cleanup` has not been run in 30 days, running now...
Removing: /Users/takahiro/Library/Caches/Homebrew/ctags--5.8_1.mojave.bottle.1.tar.gz... (143.3KB)
Removing: /Users/takahiro/Library/Caches/Homebrew/gdbm--1.18.1.mojave.bottle.1.tar.gz... (196.2KB)
Removing: /usr/local/Cellar/gettext/0.19.8.1... (1,935 files, 16.9MB)
Removing: /Users/takahiro/Library/Caches/Homebrew/heroku--7.29.0.tar.xz... (6.8MB)
Removing: /Users/takahiro/Library/Caches/Homebrew/icu4c--64.2.mojave.bottle.tar.gz... (26.1MB)
Removing: /Users/takahiro/Library/Caches/Homebrew/libunistring--0.9.10.mojave.bottle.tar.gz... (1.4MB)
Removing: /Users/takahiro/Library/Caches/Homebrew/libyaml--0.2.2.mojave.bottle.tar.gz... (100.9KB)
Removing: /Users/takahiro/Library/Caches/Homebrew/lua--5.3.5_1.mojave.bottle.tar.gz... (113.9KB)
Removing: /Users/takahiro/Library/Caches/Homebrew/mysql--8.0.17_1.mojave.bottle.tar.gz... (73.9MB)
Removing: /Users/takahiro/Library/Caches/Homebrew/node--12.9.1.mojave.bottle.tar.gz... (14.7MB)
Removing: /Users/takahiro/Library/Caches/Homebrew/openssl--1.0.2s.mojave.bottle.tar.gz... (3.7MB)
Removing: /usr/local/Cellar/openssl@1.1/1.1.1c... (7,963 files, 17.9MB)
Removing: /Users/takahiro/Library/Caches/Homebrew/openssl@1.1--1.1.1c.mojave.bottle.tar.gz... (5.3MB)
Removing: /usr/local/Cellar/python/3.7.3... (3,863 files, 59.8MB)
Removing: /usr/local/Cellar/readline/8.0.0_1... (48 files, 1.5MB)
Removing: /Users/takahiro/Library/Caches/Homebrew/readline--8.0.0_1.mojave.bottle.tar.gz... (517.7KB)
Removing: /Users/takahiro/Library/Caches/Homebrew/ruby--2.6.3.mojave.bottle.tar.gz... (9.2MB)
Removing: /usr/local/Cellar/sqlite/3.28.0... (11 files, 3.7MB)
Removing: /Users/takahiro/Library/Caches/Homebrew/xz--5.2.4.mojave.bottle.tar.gz... (373.5KB)
Removing: /Users/takahiro/Library/Caches/Homebrew/yarn--1.17.3.tar.gz... (1.2MB)
Removing: /Users/takahiro/Library/Caches/Homebrew/zsh-completions--0.30.0.tar.gz... (222.6KB)
Removing: /Users/takahiro/Library/Logs/Homebrew/heroku... (114B)
Removing: /Users/takahiro/Library/Logs/Homebrew/yarn... (100B)
Removing: /Users/takahiro/Library/Logs/Homebrew/gettext... (64B)
Removing: /Users/takahiro/Library/Logs/Homebrew/mysql... (959B)
Removing: /Users/takahiro/Library/Logs/Homebrew/heroku-node... (119B)
Removing: /Users/takahiro/Library/Logs/Homebrew/pcre2... (64B)
Removing: /Users/takahiro/Library/Logs/Homebrew/openssl@1.1... (64B)
Removing: /Users/takahiro/Library/Logs/Homebrew/node... (64B)
Removing: /Users/takahiro/Library/Logs/Homebrew/git... (64B)
Pruned 14 symbolic links and 4 directories from /usr/local
（中略）
In order to prevent conflicts with Apple's own libtool we have prepended a "g"
so, you have instead: glibtool and glibtoolize.```

最後の文の『g』が何かわからず『glibtool and glibtoolize』かなと思ったのですが、それなら特に何もしなくても良いのか？など解決策がわからなかったので、
先に「rails-erd」のみGemFileに記述しましたが
>You need to install GraphViz (http://graphviz.org/) to use this Gem.
>For more information about Ruby-Graphviz :
と出たのでRuby-Graphviz について調べ記述した
という流れでした。なので、私の認識としては
①結局GraphVizはインストールできていない
②しかし、Ruby-GraphvizがGraphvizの一種のようなものなのでrails-erdは使えた
という考えでしたが、それだと
　>"ruby-graphviz" は"rails-erd"の内部gemとして利用されているので、個別でインストールしなくても大丈夫だ
という部分と矛盾してしまうような気もするのですが何か思い違いでしょうか？

Reply:
内部gemについては理解できたと思います。
①プロジェクトルートのadd_runtime_dependencyに使いたいgemに必要なgem達が記載されている
②ただ、実際はGemFileに使いたいgemを記述してbundle installすれば必要なgem達は一緒にインストールされ、GemFile.lockも更新されるので、そのあと周辺のgemに興味が出てきたら、『GemFile.Lock』で確認できる
という認識でよろしいでしょうか？

Reply:
```
①結局GraphVizはインストールできていない
②しかし、Ruby-GraphvizがGraphvizの一種のようなものなのでrails-erdは使えた
という考えでした
```

brewの部分に関しては、多分ですがシンボリックリンクなどが上手く通っていないなどの、問題かなと。
詳細なエラーログをみないとなんとも言えないのですが、おっしゃる通りbrew経由ではGraphvizのインストールは完了していない可能性がありますね。

とはいえ、Ruby-Graphvizをインストールすることでrails-erdが使えたのであれば、よしとしましょう！

 ```
"ruby-graphviz" は"rails-erd"の内部gemとして利用されているので、個別でインストールしなくても大丈夫だ

という部分と矛盾してしまうような気もするのですが何か思い違いでしょうか？
```

こちらに関しては、基本的には個別にインストールしなくてもいいというスタンスでいいかと。
なので、もししっかりやるのであれば、

「brew doctorなどでbrewの問題点を洗い出し、改めてbrew経由でinstallできるか確かめてみる」

というようなことをしますね。

が、動くこと優先であれば一旦はこのままでもいいと思います！

Reply:
```
①プロジェクトルートのadd_runtime_dependencyに使いたいgemに必要なgem達が記載されている
②ただ、実際はGemFileに使いたいgemを記述してbundle installすれば必要なgem達は一緒にインストールされ、GemFile.lockも更新されるので、そのあと周辺のgemに興味が出てきたら、『GemFile.Lock』で確認できる
```

はい！こちらの認識で問題ないですよ！